### ISSUE TITLE
Consider dropping PHP 7.4 support or downgrading some packages

### ISSUE THREAD

**Issue by andpavlenko** [boxbilling/boxbilling#1240]
While I try to work on issue #1239, found what current `composer install` normally installed because versions is locked by composer.lock. But if try `composer update` some packages conflict with PHP 7.4. Seems like composer.json not synced with composer.lock

**composer update**

> Loading composer repositories with package information
> Updating dependencies
> Your requirements could not be resolved to an installable set of packages.
> 
> Problem 1
>     - symfony/deprecation-contracts dev-main requires php >=8.0.2 -> your php version (7.4.28) does not satisfy that requirement.
>     - symfony/deprecation-contracts 3.1.x-dev is an alias of symfony/deprecation-contracts dev-main and thus requires it to be installed too.
>     - Root composer.json requires symfony/deprecation-contracts 3.1.x-dev -> satisfiable by symfony/deprecation-contracts[3.1.x-dev (alias of dev-main)].

**php -v**

> PHP 7.4.28 (cli) (built: Feb 17 2022 16:06:35) ( NTS )

I think need to downgrade packages or bump PHP >=8.0 (best).
Whats you think, mantainers?

### BoxBilling Version

Current Master Branch

### What version of PHP are you running?

PHP 7.4

### Additional Notes

Can you invite me in Slack?

**Comment by evrifaessa** [comment]
Even though the active support for PHP 7.4 was dropped 3 months ago, it still seems to be the dominant version across the web.

But on the other hand, nearly every hosting provider should be already giving PHP 8.x support and an easy way to change PHP versions. I'd be happy to drop PHP 7.4 support if the advantages outweigh the disadvantages.

**Comment by evrifaessa** [comment]
@BenNottelling @timothygwebb any thoughts on this?

**Comment by evrifaessa** [comment]
@John-S4

**Comment by andpavlenko** [comment]
It's time to use type hinting and other modern features. ;)

**Comment by evrifaessa** [comment]
btw, I don't have a working invitation link for Slack. But it's pretty much dead anyways.

You may want to join our Discord server instead:
https://boxbilling.org/discord

**Comment by timothygwebb** [comment]
I am completely for removing PHP 7.4 support. Please also comment on my
discussion thread regarding composer and minimum stability requirement.
Links below for discussion and pull request.
https://github.com/boxbilling/boxbilling/pull/1201
https://github.com/boxbilling/boxbilling/discussions/1205

**Comment by BenNottelling** [comment]
So, I am a little reluctant to fully drop support for PHP 7.4 for a few reasons:
1) PHP 7.4 is still receiving security updates for another 8 months and is very widely used. So I would personally say that should be the minimum supported PHP version
2) Outside of basic functionality, very little of BoxBilling has been vetted for PHP 8 or 8.1 compatibility. A lot of the codebase is still very old and effectively entirely unmaintained
3) We still need to fix #1002 so on PHP 8 we don't generate massive (literal GBs) log files over time, and I'm pretty certain that functionality is straight up removed in PHP 8.1 so I'm fairly sure BB is just broken on PHP 8.1

PS: You really should join the [discord](https://boxbilling.org/discord), I have some stuff to DM you regarding BB

**Comment by andpavlenko** [comment]
1. If you want to save 7.4 compatibility, then need downgrade versions in the last stable branch.
Who need 7.4 could use last stable (beta at now) release!

2. Master branch could drop support 7.4 and start to work on refactoring legacy code to prepare next release.
In the Git branches parallel development, so one does not conflict each others.

@evrifaessa @BenNottelling @timothygwebb

**Comment by andpavlenko** [comment]
$motivation = $issue_1240_isSuccess == true ? true : false;

**Comment by evrifaessa** [comment]
I'm still okay with dropping PHP 7.4. PHP 8.0 was released in 2020 and nearly every hosting provider should have added PHP 8.x support by now.

**Comment by evrifaessa** [comment]
But yeah, we probably won't be able to run on PHP 8.1 for some time. We also should fix #1002 too.

**Comment by timothygwebb** [comment]
To fix #1002 I will be working on changing the classes to use gettype() as
soon as possible. Unless someone can get to it sooner than later.
https://www.php.net/manual/en/function.gettype.php

**Comment by andpavlenko** [comment]
If I drop 7.4 in my PR #1237 to pass the tests, community won't mind?

**Comment by evrifaessa** [comment]
> If I drop 7.4 in my PR #1237 to pass the tests, community won't mind?

Let's keep that PR on hold until we agree to drop 7.4.

**Comment by andpavlenko** [comment]
> Let's keep that PR on hold until we agree to drop 7.4.
I do't understand what is needed for this?

**Comment by timothygwebb** [comment]
It would be preferred if you set the minimum-stability in composer.json to
stable then do a composer update to get the lock file updated to a
resolvable set of dependencies before removing PHP 7.4.

**Comment by timothygwebb** [comment]
It means to make your pull request a draft until you get a resolvable set
of dependencies which can be done by applying my prior suggestion in my
last comment regarding minimum-stability.

**Comment by andpavlenko** [comment]
@timothygwebb I wrote above what composer.json contains packages not compatible with 7.4!

**Comment by evrifaessa** [comment]
> > Let's keep that PR on hold until we agree to drop 7.4.
> 
> I don't understand what is needed for this?

We don't need to pass the tests immediately. Let's just wait some more time to see if we really want to drop PHP 7.4 or keep it. Once the consensus is done, we can proceed with the PR.

**Comment by andpavlenko** [comment]
@evrifaessa 
Then downgrade packages in composer.json to 7.4. compatible.

Run `composer prohibits php 7.4`

> psr/log                        3.x-dev    requires  php (>=8.0.0)  
> symfony/deprecation-contracts  3.1.x-dev  requires  php (>=8.0.2)

**Comment by evrifaessa** [comment]
> @evrifaessa Then downgrade packages in composer.json to 7.4 compatible.
> 
> Run `composer prohibits php 7.4`
> 
> > psr/log                        3.x-dev    requires  php (>=8.0.0)
> > symfony/deprecation-contracts  3.1.x-dev  requires  php (>=8.0.2)

No. Just wait. We still don't know if we'll drop or keep PHP 7.4. There is absolutely no reed to rush.

**Comment by timothygwebb** [comment]
@evrifaessa @andpavlenko please see screenshot below my pull request did the same thing and it is now passing all checks. This is the same issue I have been working towards an agreement on regarding minimum-stability @BenNottelling and I have agreed to have minimum-stability as stable. ![Screenshot_20220330-075253.png](https://user-images.githubusercontent.com/16106877/160829310-5b027e12-ef49-4a5c-8afb-fa415f81c94b.png) @andpavlenko @evrifaessa I agree with consensus of removal of PHP 7.4 decision before proceeding with my pull request or @andpavlenko pull request. But minimum-stability should be stable as agreed upon. I will draft my pull request as well because if we decide to remove PHP 7.4 the pull request will need to be updated to resolve conflicts and update dependencies.

**Comment by andpavlenko** [comment]
@evrifaessa  I don't rush. If you want to keep 7.4 need to downgrade packages. Application don't work normally with php 7.4 at now!

@timothygwebb As I wrote at first topic, in master branch composer.json don't synced with conposer.lock. Try run to see:

> git clone https://github.com/boxbilling/boxbilling.git boxbilling-master
> cd boxbilling-master/src
> composer install
> composer update
> 
> Your requirements could not be resolved to an installable set of packages.
> 
>   Problem 1
>     - symfony/deprecation-contracts dev-main requires php >=8.0.2 -> your php version (7.4.28) does not satisfy that requirement.
>     - symfony/deprecation-contracts 3.1.x-dev is an alias of symfony/deprecation-contracts dev-main and thus requires it to be installed too.
>     - Root composer.json requires symfony/deprecation-contracts 3.1.x-dev -> satisfiable by symfony/deprecation-contracts[3.1.x-dev (alias of dev-main)].

**Comment by gOOvER** [comment]
i would keep 7.4 and drop support below 7.4. Future releases should support 8.1+

**Comment by andpavlenko** [comment]
@gOOvER Don't problem, simply resolve conflicts with support 7.4.

**Comment by timothygwebb** [comment]
> @evrifaessa  I don't rush. If you want to keep 7.4 need to downgrade packages. Application don't work normally with php 7.4 at now!
> 
> @timothygwebb As I wrote at first topic, in master branch composer.json don't synced with conposer.lock. Try run to see:
> 
> > git clone https://github.com/boxbilling/boxbilling.git boxbilling-master
> > cd boxbilling-master/src
> > composer install
> > composer update
> > 
> > Your requirements could not be resolved to an installable set of packages.
> > 
> >   Problem 1
> >     - symfony/deprecation-contracts dev-main requires php >=8.0.2 -> your php version (7.4.28) does not satisfy that requirement.
> >     - symfony/deprecation-contracts 3.1.x-dev is an alias of symfony/deprecation-contracts dev-main and thus requires it to be installed too.
> >     - Root composer.json requires symfony/deprecation-contracts 3.1.x-dev -> satisfiable by symfony/deprecation-contracts[3.1.x-dev (alias of dev-main)].

Yes and the reason why is because minimum-stability dev needs to be minimum-stability stable code example in composer.json to follow. Please reference commit fa5dbcaa8bb2f6f68d4e2bad48fe534a4e57675b https://github.com/boxbilling/boxbilling/commit/fa5dbcaa8bb2f6f68d4e2bad48fe534a4e57675b it will fix your issue. But this is not what should be done. The minimum stability from dev to stable will resolve the issue. Change in composer.json the line of code "minimum-stability": "dev", to "minimum-stability": "stable", to resolve the situation.

**Comment by timothygwebb** [comment]
> Think, you are confusing with two different issues.
> I'm use **7.4** at now, but packages don't compatible with it. ))
> 
> @timothygwebb Do you reproduce the problem as I wrote [above](https://github.com/boxbilling/boxbilling/issues/1240#issuecomment-1083063273)?
> "symfony/deprecation-contracts": "3.1.x-dev" [don't compatible](https://github.com/symfony/deprecation-contracts/blob/893fd20d9ae41a0bae2b9cbdd581ac0cf3917de3/composer.json#L18) with 7.4. Stable 3.0 also compatible only >=8.0.2
> You need to check dependencies for compatible 7.4!
> 
> This is why the tests fail.

Did you review the commit in my last comment. It shows that I understand the dependencies and their requirements. I have encountered this before and the only resolution to it is to modify the composer.lock like I did in prior mentioned commit or update the minimum stability my pull request passes all checks but doesn't need to be merged or committed because best practice solution should be to use stable dependent branches from other repos not dev.

**Comment by andpavlenko** [comment]
@timothygwebb Are your edit composer.lock manually? This should not be done at any case. All new PR will fail the test 7.4 as my #1237!

**Comment by timothygwebb** [comment]
> @timothygwebb Are you edit composer.lock manually? This should not be done at any case! All new PR will fail the test 7.4 as my #1237.

Please review in full detail

https://github.com/boxbilling/boxbilling/pull/1201

This has not been merged so it has no effect on any new pull request. But what has been committed is the minimum-stability of dev which is actually the cause of your dependency failures. As stated before please update your minimum-stability to stable.

**Comment by andpavlenko** [comment]
"minimum-stability": "stable" did not solve the problem. Because packages in composer.json is not compatible with 7.4. ((

**Comment by BenNottelling** [comment]
Let's just drop PHP 7.4 and move to stable packages in composer please

**Comment by andpavlenko** [comment]
> Let's just drop PHP 7.4 and move to stable packages in composer please

Only master branch. Especially 4.22.1.4 still compatible with 7.4. No problem.

p.s. Semver please in future releases.

**Comment by evrifaessa** [comment]
Merged. The next part should be bumping packages to their stable versions I guess.

**Comment by evrifaessa** [comment]
Did we change the minimum stability yet?

**Comment by BenNottelling** [comment]
Yes, we removed the minimum stability from composer.json which makes composer default to stable & any hard coded dev versions have been replaced

**Comment by timothygwebb** [comment]
Yes and the reason why is because minimum-stability dev needs to be
minimum-stability stable code example in composer.json to follow. Please
reference commit fa5dbcaa8bb2f6f68d4e2bad48fe534a4e57675b
https://github.com/boxbilling/boxbilling/commit/fa5dbcaa8bb2f6f68d4e2bad48fe534a4e57675b
it will fix your issue. But this is not what should be done. The minimum
stability from dev to stable will resolve the issue. Change in
composer.json the line of code "minimum-stability": "dev", to
"minimum-stability": "stable", to resolve the situation.*"minimum-stability":
"dev",**"minimum-stability": "dev",**"minimum-stability": "dev",*
